spaceship/lib/spaceship/connect_api/models/model.rb

```diff
@@ -0,0 +1,141 @@
+        end
+      end
+
+      #
+      # Example:
+      # { "minOsVersion" => "min_os_version" }
+      #
+      # Creates attr_write and attr_reader for :min_os_version
+      # Creates alias for :minOsVersion to :min_os_version
+      #
+      def attr_mapping(attr_map)
+        attr_map.each do |key, value|
+          # Actual
+          reader = value.to_sym
+          writer = "#{value}=".to_sym
+
+          has_reader = instance_methods.include?(reader)
+          has_writer = instance_methods.include?(writer)
+
+          send(:attr_reader, value) unless has_reader
+          send(:attr_writer, value) unless has_writer
+          
+          # Alias
+          key_reader = key.to_sym
+          key_writer = "#{key}=".to_sym
+
+          # Alias the API response name to attribute name
+          alias_method key_reader, reader
+          alias_method key_writer, writer
+        end
+      end
+    end
+
+    module Models
+      class << self
+        attr_accessor :types
+        attr_accessor :types_cache
+      end
+
+      def self.parse(json)
+        data = json["data"]
+        raise "No data" unless data
+
+        included = json["included"] || []
+
+        if data.is_a?(Hash)
+          inflate_model(model_data, included)
+        elsif data.is_a?(Array)
+          return data.map do |model_data|
+            inflate_model(model_data, included)
+          end
+        else
+          puts "something else"
+        end
+      end
+
+      def self.find_class(model_data)
+        # Initialize cache
+        @types_cache ||= {}
+
+        # Find class in cache
+        type_string = model_data["type"] 
+        type_class = @types_cache[type_string]
+        return type_class if type_class
+
+        # Find class in array
+        type_class = @types.find do |type|
+          type.type == type_string
+        end
+
+        # Cache and return class
+        @types_cache[type_string] = type_class
+        return type_class
+      end
+
+      def self.inflate_model(model_data, included)
+        # Find class
+        type_class = find_class(model_data)
+        raise "No type class found for #{model_data["type"]}" unless type_class
+
+        # Get id and attributes needed for inflating
+        id = model_data["id"]
+        attributes = model_data["attributes"]
+
+        # Instantiate object and inflate relationships
+        type_instance = type_class.new(id, attributes)
+        type_instance = inflate_model_relationships(type_instance, model_data, included)
+
+        return type_instance
+      end
+
+      def self.inflate_model_relationships(type_instance, model_data, included)
+        # Relationship attributes to set
+        attributes = {}
+
+        # 1. Iterate over relations
+        # 2. Find id and type
+        # 3. Find matching id and type in included
+        # 4. Inflate matching data and set in attributes
+        relationships = model_data["relationships"] || []
```

Comment:
it looks like `model_data` is only needed to get at the `relationships`. Lets change this method so that it accepts `relationships` instead of model data and promote that lookup outside of this method.